### PR TITLE
CLI improvements proposal: positional argument

### LINKED ISSUE #56: Migration tool CLI redesign
#### Summary

`lorawan-stack-migrate` tool could use a different approach, mainly the flag system.

#### Why do we need this?

The migration has been growing and the flag system is getting out of hand.

The way we implement flags currently does not support adding flags specific to a command (`applications`, `device`) or sources. This creates a lot of fuzz in the flags we present to the user that are not relevant for the current use case and it's only going to get worse:
image

An example of this is the flag `app-id` is not needed when using `applications` command, because the `app-id` is supplied in the command itself.

Other than that `firefly` source wouldn't need `applications` command as much as `organizations` as to what I've seen that is used more often to organize devices on the platform. Because of that being able to define commands per source would be beneficial as well.

#### What is already there? What do you see now?

We have a working CLI.

#### What is missing? What do you want to see?

Flexibility and structure of the CLI.

#### Environment

n/a

#### How do you propose to implement this?

Use a well developed CLI library like [`spf13/cobra`](https://github.com/spf13/cobra) and rethink the way we implement stuff.

#### How do you propose to test this?

Test migrating devices with new CLI.

#### Can you do this yourself and submit a Pull Request?

I can implement this myself, but it would require help in design from a more experienced developer.

### PR DESCRIPTION
**Pull request by happyRip**
#### Summary

Closes #56  

#### Changes

- Add an optional positional `source` argument (`pkg/arg`)
 
- Print only source-related flags with those commands
  - Removed prefixes from `ttnv3` flags; **to be followed with other sources**

- Split  `application` and `devices` commands FlagSets

- Improved `ttnv3` config by utilising flags that save value to variables directly

#### Testing

local testing

- [x] The Things Stack
- [x] Chirpstack v3
- [ ] ttnv2

##### Regressions

I do not expect any.

#### Notes for Reviewers

- If we were planning to add more commands in the future, FlagSet splitting could be upgraded to a more sophisticated solution. Currently it's pretty simple.

- It was suggested to use generated commands like `tts-devices` ETC, but I wanted to propose this solution which seems to work well and is pretty simple

- When `pflag` is used to set ActiveSource the value is set after initialisation, so the flags cannot be separated easily (during [`FlagSet()`](https://github.com/TheThingsNetwork/lorawan-stack-migrate/pull/65/files#diff-54e8061fb2925948c62f36b19e08785ec1fb90b349cfe48c73239f4cca8c6ef5R82-R99) function call)

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [ ] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.

### CODE REVIEW

File: cmd/ttn-lw-migrate/main.go [hunk 1]
```diff
@@ -21,6 +21,8 @@ import (
 	_ "go.thethings.network/lorawan-stack-migrate/pkg/source/ttnv2"      // TTNv2 source
 	_ "go.thethings.network/lorawan-stack-migrate/pkg/source/ttnv3"      // TTS source
 
+	_ "go.thethings.network/lorawan-stack-migrate/pkg/arg" // Init after initializing sources
```
Comment (adriansmares) [inline cmd/ttn-lw-migrate/main.go:24]: So we are doing this, such that `arg.init` runs before the `pflag` initialization ?

I am not fully against this idea, but keep in mind that you can edit the arguments of a `cobra.Command` using [`SetArgs`](https://pkg.go.dev/github.com/spf13/cobra#Command.SetArgs), perhaps in the `init` of `cmd/root`.

This in turns allows you basically intercept the parameters between `os.Args` and the actual usage, thus avoiding the clever-but-complicated solution.

Comment (happyRip) [reply cmd/ttn-lw-migrate/main.go:24]: Using [`SetArgs`](https://pkg.go.dev/github.com/spf13/cobra#Command.SetArgs) instead of removing the field from `os.Args` seems like a great idea. The problem is that `application` command's `init` function is always executed first. The argument has to be interpreted before [`source.FlagSet`](https://github.com/TheThingsNetwork/lorawan-stack-migrate/blob/master/pkg/source/source.go#L79) is called for it to work.

Comment (adriansmares) [reply cmd/ttn-lw-migrate/main.go]: I see - `ApplicationFlagSet` and `DevicesFlagSet` depend on the active source being set in order to have the intended behavior. Instead of depending on `init` order, let's just move the initialization after the active source has been set.

Let me know what you think of f173c6c6367566ad30050998db8c2c2ef0486aef.

File: pkg/source/ttnv3/config.go [hunk 2]
```diff
@@ -15,121 +14,197 @@ import (
...
+	appID string
+
+	serverConfig serverConfig
 
 	deleteSourceDevice,
 	dryRun,
 	noSession bool
 }
 
+var config Config
```
Comment (adriansmares) [inline pkg/source/ttnv3/config.go:65]: So if my understanding is correct here, you have added this global state in order to work around the fact that you cannot pass state between `flagSet()` and `NewSource`.

Historically (in other sources) it is indeed the case that the idea was that these two are decoupled - you created the flags with some kind of `flagSet()`, then read the flags from the flag set in `NewSource`. But please note that nothings stops you from actually having state pass between the two.

You can have `flagSet` also return a `*Config`, and then the `NewSource` function can be built using this `*Config`:
```go
func createNewSource(config *Config) func(context.Context, *pflag.FlagSet) (source.Source, error) {
	return func(ctx context.Context, _ *pflag.FlagSet) (source.Source, error) {
		return Source{
			ctx:    ctx,
			config: config,
		}, nil
	}
}
```

I know that `getConfig` also does some parsing but you can consider moving that to this `newSource` factory function.

You can consider even dropping the `*pflag.FlagSet` from the `NewSource` function interface - this is a bit more work but I encourage it.

File: cmd/ttn-lw-migrate/main.go (hunk 1 above)
Comment (happyRip) [reply cmd/ttn-lw-migrate/main.go]: It looks great

File: pkg/source/ttnv3/config/config.go [hunk 3]
```diff
@@ -56,61 +56,66 @@ func (c *serverConfig) anyFieldEmpty() error {
...
 	var (
-		config = &Config{}
-		flags  = &pflag.FlagSet{}
+		config  = &Config{}
+		devices = &pflag.FlagSet{}
+		shared  = &pflag.FlagSet{}
 	)
 
-	flags.StringVar(&config.AppID,
+	devices.StringVar(&config.AppID,
```
Comment (happyRip) [inline pkg/source/ttnv3/config/config.go:80]: The `--app-id` flag was removed from `application` command of `ttnv{2,3}` sources since there was no use case for it.
Example:
```bash
ttn-lw-migrate ttnv3 application --app-id 'my-app'
```
If the tool was called like this without an argument it would expect to receive Application IDs from `stdin` so the flag would be ignored and the tool would 'hang unexpectedly' (waiting).

Since https://github.com/TheThingsNetwork/lorawan-stack-migrate/commit/f33a68b28ac0b0bb2fd987b411689a6b1374b763 the argument always overrides the flag. This combination makes the flag redundant or even confusing for the `application` command in those sources.

File: CHANGELOG.md [hunk 4]
```diff
@@ -8,24 +8,31 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
...
-### Fixed
+- depreciated `--source` flag
 
-- `app-id` flag/variable is no longer needed when calling `application` command with `ttnv2` and `ttnv3` sources
-- `application` command no longer panics when called with a different `app-id` argument than set with flag/variable
+### Fixed
 
 ### Removed
 
+- `--app-id` flag for `application` command
```
Comment (adriansmares) [inline CHANGELOG.md:21]: Dot at the end. Same below.

File: pkg/source/ttnv3/config/config.go [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+			setCustomCA(c.caPath)
+		}
+	}
+
+	c.ServerConfig.applyDefaults()
+	if err := c.ServerConfig.anyFieldEmpty(); err != nil {
+		return err
+	}
+
+	// deleteSourceDevice not allowed when dryRun
```
Comment (adriansmares) [inline pkg/source/ttnv3/config/config.go:178]: `DeleteSourceDevice is not allowed during a dry run`.

File: pkg/source/ttnv3/util.go [hunk 6]
```diff
@@ -59,7 +59,6 @@ func clearDeviceSession(dev *ttnpb.EndDevice) error {
 	return dev.SetFields(nil,
 		"activated_at",
 		"mac_state",
-		"last_dev_status_f_cnt_up",
```
Comment (adriansmares) [inline pkg/source/ttnv3/util.go:62]: Why this change ?

**Comment by adriansmares** [comment]
> Is this flag magic really necessary? Why don't we have commands per source with their own flag sets?

@happyRip I think this actually may be better than our current hack. We would need to generate dynamically an `application` / `device` sub command per source, but I think it would indeed work better here.

File: pkg/source/ttnv3/util.go (hunk 6 above)
Comment (happyRip) [reply pkg/source/ttnv3/util.go:62]: `last_dev_status_f_cnt_up` is a part of `mac_state`. Keeping this line resulted in an error:
image

**Comment by adriansmares** [comment]
What is the status here - is this blocked on anything ? Does it need any technical decision ?

**Comment by happyRip** [comment]
[Subcommands approach](https://github.com/TheThingsNetwork/lorawan-stack-migrate/pull/71) was merged, so this PR can be closed.
